### PR TITLE
Instrumentation for jboss-logmanager getMdcCopy() (#6111)

### LINKED ISSUE #6111: jboss-logmanager getMdcCopy() doesn't contain trace information
**Is your feature request related to a problem? Please describe.**
The library jboss-logmanager contains since version 1.3 a new method getMdcCopy(). This one returns a map with all MDC attributes. In order to be able to log span/trace IDs for example in a JSON log, the return value should contain the values like getMdc(attribute) already supports.

**Describe the solution you'd like**
Extended the instrumentation to return a map that includes those attributes if they are now already set.

**Describe alternatives you've considered**
It would be possible to create a custom instrumentation, still I think this would be useful for everyone else as well. 

**Additional context**

The version 1.3 has been release in Apr 2012, almost 10 years ago, and the API is available in the current versions as well. 

I'll prepare a PR shortly later today.

This would allow to log the trace/span information in my setup where a Quarkus application logs JSON. 
Before that change the MDC would be empty, now it contains the tracing information: 

> {"timestamp":"2022-05-28T14:33:44.197+02:00","sequence":8441,"loggerClassName":"org.jboss.logging.Logger","loggerName":"org.keycloak.events","level":"WARN","message":"type=LOGIN_ERROR, realmId=84f7eab5-4af2-40a3-ab44-783b76f5919a, c
lientId=null, userId=null, ipAddress=127.0.0.1, error=expired_code, restart_after_timeout=true, authSessionParentId=37f81c2c-2db1-4b7c-9a1c-6d0d7b2fd552, authSessionTabId=Xp6HO7eT8V4","threadName":"executor-thread-0","threadId":24,"mdc":{"trace_flags":"aa4a99929037cee4a35fb8666b924c04","trace_id":"aa4a99929037cee4a35fb8666b924c04","span_id":"aa4a99929037cee4a35fb8666b924c04"},"ndc":"","hostName":"tiny","processName":"QuarkusEntryPoint","processId":17268}

### PR DESCRIPTION
**Pull request by ahus1**
This allows, for example, in JSON logs to have the MDC to contain span and trace information.

Both the unit test and a manual integration test with a Quarkus application worked as expected for me. 

I wondered if I should add a new instrumentation module as the new method was added in 1.3. As that was release in Apr 2012 and other methods stay identical, I decided not to do that unless someone else requires this. 

Please advise if you'd like to see a different approach, and I'll update this PR.

Closes #6111

### CODE REVIEW

File: instrumentation/jboss-logmanager/jboss-logmanager-mdc-1.1/javaagent/build.gradle.kts [hunk 1]
```diff
@@ -12,5 +12,5 @@ muzzle {
 }
 
 dependencies {
-  library("org.jboss.logmanager:jboss-logmanager:1.1.0.GA")
+  library("org.jboss.logmanager:jboss-logmanager:1.3.0.Final")
```
Comment (laurit) [inline instrumentation/jboss-logmanager/jboss-logmanager-mdc-1.1/javaagent/build.gradle.kts:15]: If possible we compile and test against the earliest version supported. If introducing a new instrumentation feature makes the instrumentation incompatible then the version number should be bumped everywhere, in module name, package name, docs etc.
As far as I understand you are only calling `getMdcCopy` from test, which is using groovy, so you can probably change this back to 1.1.0. In test you could check whether this method is present before calling it. Something like https://github.com/open-telemetry/opentelemetry-java-instrumentation/blob/efd40f8cdf73f33b8a055359d7fe8bed5afaf54c/instrumentation/netty/netty-3.8/javaagent/src/test/groovy/Netty38ClientTest.groovy#L33 When you run tests with `-PtestLatestDeps=true` tests will run against latest version of `jboss-logmanager` and thus test your changes too. Alternatively it is possible to define `testSets` using `org.unbroken-dome.test-sets` plugin so you can test with multiple versions each of which can have its own tests, see https://github.com/open-telemetry/opentelemetry-java-instrumentation/blob/main/instrumentation/jsf/jsf-mojarra-1.2/javaagent/build.gradle.kts Getting test sets to work exactly as you want can be tricky.

File: instrumentation/jboss-logmanager/jboss-logmanager-mdc-1.1/javaagent/src/test/groovy/JbossLogmanagerMdcTest.groovy [hunk 2]
```diff
@@ -80,6 +80,10 @@ class JbossLogmanagerMdcTest extends AgentInstrumentationSpecification {
     logRecords.get(2).getMdc("trace_id") == span1.spanContext.traceId
     logRecords.get(2).getMdc("span_id") == span1.spanContext.spanId
     logRecords.get(2).getMdc("trace_flags") == "01"
+    def mdcCopy2 = logRecords.get(2).getMdcCopy()
+    mdcCopy2.get("trace_id") != null
```
Comment (laurit) [inline instrumentation/jboss-logmanager/jboss-logmanager-mdc-1.1/javaagent/src/test/groovy/JbossLogmanagerMdcTest.groovy:84]: would it be possible to validate against expected value?

Comment (ahus1) [reply instrumentation/jboss-logmanager/jboss-logmanager-mdc-1.1/javaagent/src/test/groovy/JbossLogmanagerMdcTest.groovy]: Updated it to test against the expected value.

File: instrumentation/jboss-logmanager/jboss-logmanager-mdc-1.1/javaagent/build.gradle.kts (hunk 1 above)
Comment (ahus1) [reply instrumentation/jboss-logmanager/jboss-logmanager-mdc-1.1/javaagent/build.gradle.kts]: Thank for the advice, I re-wrote the test, and also reverted the dependency version. 

I found out the hard way that I must not use an `if` statement within the `then` block. I have to admit I've never used Spock to write tests, so I'm happy to hear a better way to do this.

Comment (laurit) [reply instrumentation/jboss-logmanager/jboss-logmanager-mdc-1.1/javaagent/build.gradle.kts]: Try
```groovy
if (hasGetMdcCopy) {
  assert logRecords.get(2).getMdcCopy().get("trace_id") == span1.spanContext.traceId
  assert logRecords.get(2).getMdcCopy().get("span_id") == span1.spanContext.spanId
  assert logRecords.get(2).getMdcCopy().get("trace_flags") == "01"
}
```

Comment (ahus1) [reply instrumentation/jboss-logmanager/jboss-logmanager-mdc-1.1/javaagent/build.gradle.kts]: Thanks for pointing me to the `assert` keyword that seems to be necessary within an if block. Without that it evaluated always to true. Ready for a (final?) review.

**Review by laurit — APPROVED** [review]

**Review by trask — COMMENTED** [review]
hi @ahus1!

it looks like `getMdcCopy()` does two things:

* makes a copy of the thread local MDC values and stores that into the LogRecord, so that the LogRecord can be passed to another thread for async logging
* returns a copy of that copy

I think the first point above is the primary use case (similar to log4j's `getMDCCopy()` which returns `void`), and so I'm not sure if this PR solves that case, or if I'm missing the use case for doing something with the return value from `getMdcCopy()`.

**Comment by ahus1** [comment]
Hi @trask - thanks for reviewing this PR. 

The `getMdcCopy()` is copying the thread-local MDC values, and it is also (in contrast to log4j) the only way to retrieve that map. This implementation is symmetric to the instrumentation of the get `getMdc(key)` call: it first runs the original method (which will do the copy), and then updates the result returned to the caller.  

The `getMdcCopy()`is the method used when logging the record as JSON - the previous instrumentation will not log the MDC contents in the JSON output; see #6111.

For me, this is the way to augment the MDC map to contain the trace information, which AFAIK is the only way to display the trace information when logging the record as JSON. The map is retrieved here from the record: https://github.com/jboss-logging/jboss-logmanager/blob/c6d88b929aaab6fbc815c249bade8891661853f9/src/main/java/org/jboss/logmanager/formatters/StructuredFormatter.java#L215-L224

I tested this successfully when building the agent and running with in my case a Quarkus based app running in standard, non-native mode.

I hope this answers your questions about the intent of the changes, and why I chose this approach.

**Review by trask — APPROVED** [review]
> it is also (in contrast to log4j) the only way to retrieve that map

oh yes, this is what I missed, thx!

**Pull request merged**